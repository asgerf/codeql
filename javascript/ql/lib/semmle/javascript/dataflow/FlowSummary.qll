/** Provides classes and predicates for defining flow summaries. */

private import javascript
private import semmle.javascript.dataflow.internal.sharedlib.FlowSummaryImpl as Impl
private import semmle.javascript.dataflow.internal.FlowSummaryPrivate
private import semmle.javascript.dataflow.internal.sharedlib.DataFlowImplCommon as DataFlowImplCommon
private import semmle.javascript.dataflow.internal.DataFlowPrivate

/** A callable with a flow summary, identified by a unique string. */
abstract class SummarizedCallable extends LibraryCallable, Impl::Public::SummarizedCallable {
  bindingset[this]
  SummarizedCallable() { any() }

  // TODO: rename 'propagatesFlowExt' and/or override 'propagatesFlow' directly
  /**
   * Holds if data may flow from `input` to `output` through this callable.
   *
   * `preservesValue` indicates whether this is a value-preserving step or a taint-step.
   */
  pragma[nomagic]
  predicate propagatesFlowExt(string input, string output, boolean preservesValue) { none() }

  override predicate propagatesFlow(
    string input, string output, boolean preservesValue, string model
  ) {
    this.propagatesFlowExt(input, output, preservesValue) and model = this
  }

  /**
   * Gets the synthesized parameter that results from an input specification
   * that starts with `Argument[s]` for this library callable.
   */
  DataFlow::ParameterNode getParameter(string s) {
    exists(ParameterPosition pos |
      DataFlowImplCommon::parameterNode(result, MkLibraryCallable(this), pos) and
      s = encodeParameterPosition(pos)
    )
  }
}
