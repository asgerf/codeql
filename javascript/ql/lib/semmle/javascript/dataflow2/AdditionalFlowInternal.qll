private import javascript
private import semmle.javascript.dataflow2.DataFlow as DataFlow2
private import semmle.javascript.dataflow2.DataFlowImplSpecific
private import semmle.javascript.dataflow.internal.DataFlowNode

/**
 * Gets a data-flow node synthesized using `AdditionalFlowInternal#needsSynthesizedNode`.
 */
DataFlow::Node getSynthesizedNode(AstNode node, string tag) {
  result = TGenericSynthesizedNode(node, tag, _)
}

/**
 * An extension to `AdditionalFlowStep` with additional internal-only predicates.
 */
class AdditionalFlowInternal extends DataFlow::AdditionalFlowStep {
  /**
   * Holds if a data-flow node should be synthesized for the pair `(node, tag)`.
   *
   * The node can be obtained using `getSynthesizedNode(node, tag)`.
   *
   * `container` will be seen as the node's enclosing container.
   */
  predicate needsSynthesizedNode(AstNode node, string tag, StmtContainer container) { none() }

  /**
   * Holds if `node` should only permit flow of values stored in `contents`.
   */
  predicate expectsContent(DataFlow::Node node, DataFlow2::ContentSet contents) { none() }

  /**
   * Holds if `node` should not permit flow of values stored in `contents`.
   */
  predicate clearsContent(DataFlow::Node node, DataFlow2::ContentSet contents) { none() }
}
