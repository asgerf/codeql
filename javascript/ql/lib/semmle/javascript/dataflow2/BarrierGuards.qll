/**
 * A copy of the barrier guard logic from `Configuration.qll` in the JS data flow library.
 *
 * This version considers all barrier guards to be relevant.
 */

private import javascript
private import semmle.javascript.dataflow.internal.AccessPaths

signature predicate isBarrierGuardSig(DataFlow::BarrierGuardNode node);

module MakeBarrierGuards<isBarrierGuardSig/1 isBarrierGuard> {
  final private class FinalNode = DataFlow::Node;

  abstract private class BarrierGuard extends FinalNode {
    /**
     * Holds if data flow node `guard` acts as a barrier for data flow.
     *
     * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
     */
    abstract predicate blocksExpr(boolean outcome, Expr test, string label);
  }

  class ExplicitBarrierGuard extends BarrierGuard instanceof DataFlow::BarrierGuardNode {
    ExplicitBarrierGuard() {
      isBarrierGuard(this)
      or
      this instanceof DataFlow::AdditionalBarrierGuardNode
    }

    override predicate blocksExpr(boolean outcome, Expr test, string label) {
      super.blocks(outcome, test, label)
      or
      super.blocks(outcome, test) and label = ""
    }
  }

  /**
   * Holds if data flow node `guard` acts as a barrier for data flow.
   *
   * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
   */
  pragma[nomagic]
  private predicate barrierGuardBlocksExpr(
    BarrierGuard guard, boolean outcome, Expr test, string label
  ) {
    guard.blocksExpr(outcome, test, label)
  }

  /**
   * Holds if `guard` may block the flow of a value reachable through exploratory flow.
   */
  pragma[nomagic]
  private predicate barrierGuardIsRelevant(BarrierGuard guard) {
    exists(Expr e |
      barrierGuardBlocksExpr(guard, _, e, _)
      // All guards are considered relevant (this is the difference from the main JS lib)
      // isRelevantForward(e.flow(), _)
    )
  }

  /**
   * Holds if data flow node `guard` acts as a barrier for data flow due to aliasing through
   * an access path.
   *
   * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
   */
  pragma[nomagic]
  private predicate barrierGuardBlocksAccessPath(
    BarrierGuard guard, boolean outcome, AccessPath ap, string label
  ) {
    barrierGuardIsRelevant(guard) and
    barrierGuardBlocksExpr(guard, outcome, ap.getAnInstance(), label)
  }

  /**
   * Holds if there exists an input variable of `ref` that blocks the label `label`.
   *
   * This predicate is outlined to give the optimizer a hint about the join ordering.
   */
  pragma[nomagic]
  private predicate barrierGuardBlocksSsaRefinement(
    BarrierGuard guard, boolean outcome, SsaRefinementNode ref, string label
  ) {
    barrierGuardIsRelevant(guard) and
    guard.getEnclosingExpr() = ref.getGuard().getTest() and
    forex(SsaVariable input | input = ref.getAnInput() |
      barrierGuardBlocksExpr(guard, outcome, input.getAUse(), label)
    )
  }

  /**
   * Holds if the result of `guard` is used in the branching condition `cond`.
   *
   * `outcome` is bound to the outcome of `cond` for join-ordering purposes.
   */
  pragma[nomagic]
  private predicate barrierGuardUsedInCondition(
    BarrierGuard guard, ConditionGuardNode cond, boolean outcome
  ) {
    barrierGuardIsRelevant(guard) and
    outcome = cond.getOutcome() and
    (
      cond.getTest() = guard.getEnclosingExpr()
      or
      cond.getTest().flow().getImmediatePredecessor+() = guard
    )
  }

  /**
   * Holds if data flow node `nd` acts as a barrier for data flow, possibly due to aliasing
   * through an access path.
   *
   * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
   */
  pragma[nomagic]
  private predicate barrierGuardBlocksNode(BarrierGuard guard, DataFlow::Node nd, string label) {
    // 1) `nd` is a use of a refinement node that blocks its input variable
    exists(SsaRefinementNode ref, boolean outcome |
      nd = DataFlow::ssaDefinitionNode(ref) and
      outcome = ref.getGuard().(ConditionGuardNode).getOutcome() and
      barrierGuardBlocksSsaRefinement(guard, outcome, ref, label)
    )
    or
    // 2) `nd` is an instance of an access path `p`, and dominated by a barrier for `p`
    exists(AccessPath p, BasicBlock bb, ConditionGuardNode cond, boolean outcome |
      nd = DataFlow::valueNode(p.getAnInstanceIn(bb)) and
      barrierGuardUsedInCondition(guard, cond, outcome) and
      barrierGuardBlocksAccessPath(guard, outcome, p, label) and
      cond.dominates(bb)
    )
  }

  /**
   * Gets a logical `and` expression, or parenthesized expression, that contains `guard`.
   */
  private Expr getALogicalAndParent(BarrierGuard guard) {
    barrierGuardIsRelevant(guard) and result = guard.asExpr()
    or
    result.(LogAndExpr).getAnOperand() = getALogicalAndParent(guard)
    or
    result.getUnderlyingValue() = getALogicalAndParent(guard)
  }

  /**
   * Gets a logical `or` expression, or parenthesized expression, that contains `guard`.
   */
  private Expr getALogicalOrParent(BarrierGuard guard) {
    barrierGuardIsRelevant(guard) and result = guard.asExpr()
    or
    result.(LogOrExpr).getAnOperand() = getALogicalOrParent(guard)
    or
    result.getUnderlyingValue() = getALogicalOrParent(guard)
  }

  final private class FinalFunction = Function;

  /**
   * A function that returns the result of a barrier guard.
   */
  private class BarrierGuardFunction extends FinalFunction {
    DataFlow::ParameterNode sanitizedParameter;
    BarrierGuard guard;
    boolean guardOutcome;
    string label;
    int paramIndex;

    BarrierGuardFunction() {
      barrierGuardIsRelevant(guard) and
      exists(Expr e |
        exists(Expr returnExpr |
          returnExpr = guard.asExpr()
          or
          // ad hoc support for conjunctions:
          getALogicalAndParent(guard) = returnExpr and guardOutcome = true
          or
          // ad hoc support for disjunctions:
          getALogicalOrParent(guard) = returnExpr and guardOutcome = false
        |
          exists(SsaExplicitDefinition ssa |
            ssa.getDef().getSource() = returnExpr and
            ssa.getVariable().getAUse() = this.getAReturnedExpr()
          )
          or
          returnExpr = this.getAReturnedExpr()
        ) and
        sanitizedParameter.flowsToExpr(e) and
        barrierGuardBlocksExpr(guard, guardOutcome, e, label)
      ) and
      sanitizedParameter.getParameter() = this.getParameter(paramIndex)
    }

    /**
     * Holds if this function sanitizes argument `e` of call `call`, provided the call evaluates to `outcome`.
     */
    predicate isBarrierCall(DataFlow::CallNode call, Expr e, boolean outcome, string lbl) {
      exists(DataFlow::Node arg |
        DataFlow::argumentPassingStep(pragma[only_bind_into](call), pragma[only_bind_into](arg),
          pragma[only_bind_into](this), pragma[only_bind_into](sanitizedParameter)) and
        arg.asExpr() = e and
        arg = call.getArgument(paramIndex) and
        outcome = guardOutcome and
        lbl = label
      )
    }
  }

  /**
   * A call that sanitizes an argument.
   */
  private class AdditionalBarrierGuardCall extends BarrierGuard instanceof DataFlow::CallNode {
    BarrierGuardFunction f;

    AdditionalBarrierGuardCall() { f.isBarrierCall(this, _, _, _) }

    override predicate blocksExpr(boolean outcome, Expr e, string label) {
      f.isBarrierCall(this, e, outcome, label)
    }
  }

  /**
   * A sanitizer where an inner sanitizer is compared against a boolean.
   * E.g. (assuming `sanitizes(e)` is an existing sanitizer):
   * ```javascript
   * if (sanitizes(e) === true) {
   *  // e is sanitized
   * }
   * ```
   */
  private class CallAgainstEqualityCheck extends BarrierGuard {
    BarrierGuard prev;
    boolean polarity;

    CallAgainstEqualityCheck() {
      prev instanceof DataFlow::CallNode and
      exists(EqualityTest test, BooleanLiteral bool |
        this.asExpr() = test and
        test.hasOperands(prev.asExpr(), bool) and
        polarity = test.getPolarity().booleanXor(bool.getBoolValue())
      )
    }

    override predicate blocksExpr(boolean outcome, Expr e, string lbl) {
      exists(boolean prevOutcome |
        barrierGuardBlocksExpr(prev, prevOutcome, e, lbl) and
        outcome = prevOutcome.booleanXor(polarity)
      )
    }
  }

  pragma[nomagic]
  predicate barrierGuardBlocksNode(DataFlow::Node node, string label) {
    barrierGuardBlocksNode(_, node, label)
  }

  pragma[nomagic]
  predicate barrierGuardBlocksNode(DataFlow::Node node) { barrierGuardBlocksNode(_, node, "") }

  pragma[nomagic]
  predicate barrierGuardBlocksNodeIncludeHeuristicCheck(DataFlow::Node nd, string label) {
    exists(BarrierGuard guard | barrierGuardBlocksNode(guard, nd, label))
  }
}

module MakeSanitizerGuards<isBarrierGuardSig/1 isBarrierGuard> {
  private predicate isBarrierGuard2(DataFlow::BarrierGuardNode node) {
    isBarrierGuard(node)
    or
    node instanceof TaintTracking::AdditionalSanitizerGuardNode
  }

  import MakeBarrierGuards<isBarrierGuard2/1>
}

private predicate empty(DataFlow::BarrierGuardNode node) { none() }

module DefaultSanitizerGuards = MakeSanitizerGuards<empty/1>;

/** There are no barrier guards by default */
module DefaultBarrierGuards {
  predicate barrierGuardBlocksNode(DataFlow::Node nd, string label) { none() }

  predicate barrierGuardBlocksNode(DataFlow::Node nd) { none() }
}
