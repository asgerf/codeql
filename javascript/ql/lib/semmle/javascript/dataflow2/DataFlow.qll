private import codeql.dataflow.DataFlow
private import DataFlowArg
import DataFlowMake<JSDataFlow>
import DataFlowImplSpecific::Public
