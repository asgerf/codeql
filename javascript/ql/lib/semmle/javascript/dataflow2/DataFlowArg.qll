private import DataFlowImplSpecific
private import codeql.dataflow.DataFlow
private import codeql.dataflow.TaintTracking as TT

module JSDataFlow implements InputSig {
  import Private
  import Public

  predicate typeStrongerThan(DataFlowType t1, DataFlowType t2) { none() }

  predicate neverSkipInPathGraph = Private::neverSkipInPathGraph/1;

  predicate accessPathLimit = Private::accessPathLimit/0;
}

module JSTaintFlow implements TT::InputSig<JSDataFlow> {
  private import Private
  private import Public

  /**
   * Holds if `node` should be a sanitizer in all global taint flow configurations
   * but not in local taint.
   */
  predicate defaultTaintSanitizer(Node node) { none() }

  /**
   * Holds if default `TaintTracking::Configuration`s should allow implicit reads
   * of `c` at sinks and inputs to additional taint steps.
   */
  bindingset[node]
  predicate defaultImplicitTaintRead(Node node, ContentSet c) { none() }

  predicate defaultAdditionalTaintStep = Private::defaultAdditionalTaintStep/2;
}
