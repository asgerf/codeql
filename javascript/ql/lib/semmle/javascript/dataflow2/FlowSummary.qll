/** Provides classes and predicates for defining flow summaries. */

private import javascript
private import FlowSummaryImpl as Impl
private import DataFlowImplCommon as DataFlowImplCommon
private import DataFlowImplSpecific::Private
private import DataFlowImplSpecific::Public
private import FlowSummaryImplSpecific

class SummaryComponent = Impl::Public::SummaryComponent;

/** Provides predicates for constructing summary components. */
module SummaryComponent {
  private import Impl::Public::SummaryComponent as SC

  predicate parameter = SC::parameter/1;

  predicate argument = SC::argument/1;

  predicate content = SC::content/1;

  predicate withoutContent = SC::withoutContent/1;

  predicate withContent = SC::withContent/1;

  class SyntheticGlobal = SC::SyntheticGlobal;

  /** Gets a summary component that represents a receiver. */
  SummaryComponent receiver() { result = argument(-1) }

  /** Gets a summary component that represents an element in a collection at an unknown index. */
  SummaryComponent arrayElement() {
    result = SC::content(ContentSet::singleton(DataFlow::PseudoProperties::arrayElement()))
  }

  /** Gets a summary component that represents the return value of a call. */
  SummaryComponent return() { result = SC::return(MkNormalReturnKind()) }

  /** Gets a summary component that represents the exception thrown from a call. */
  SummaryComponent exceptionalReturn() { result = SC::return(MkExceptionalReturnKind()) }
}

class SummaryComponentStack = Impl::Public::SummaryComponentStack;

/** Provides predicates for constructing stacks of summary components. */
module SummaryComponentStack {
  private import Impl::Public::SummaryComponentStack as SCS

  predicate singleton = SCS::singleton/1;

  predicate push = SCS::push/2;

  predicate argument = SCS::argument/1;

  /** Gets a singleton stack representing a receiver. */
  SummaryComponentStack receiver() { result = singleton(SummaryComponent::receiver()) }

  /** Gets a singleton stack representing the return value of a call. */
  SummaryComponentStack return() { result = singleton(SummaryComponent::return()) }

  /** Gets a singleton stack representing the exception thrown from a call. */
  SummaryComponentStack exceptionalReturn() {
    result = singleton(SummaryComponent::exceptionalReturn())
  }
}

/** A callable with a flow summary, identified by a unique string. */
abstract class SummarizedCallable extends LibraryCallable, Impl::Public::SummarizedCallable {
  bindingset[this]
  SummarizedCallable() { any() }

  /**
   * Same as
   *
   * ```ql
   * propagatesFlow(
   *   SummaryComponentStack input, SummaryComponentStack output, boolean preservesValue
   * )
   * ```
   *
   * but uses an external (string) representation of the input and output stacks.
   */
  pragma[nomagic]
  predicate propagatesFlowExt(string input, string output, boolean preservesValue) { none() }

  /**
   * Gets the synthesized parameter that results from an input specification
   * that starts with `Argument[s]` for this library callable.
   */
  DataFlow::ParameterNode getParameter(string s) {
    exists(ParameterPosition pos |
      DataFlowImplCommon::parameterNode(result, MkLibraryCallable(this), pos) and
      s = getParameterPosition(pos)
    )
  }
}

class RequiredSummaryComponentStack = Impl::Public::RequiredSummaryComponentStack;
