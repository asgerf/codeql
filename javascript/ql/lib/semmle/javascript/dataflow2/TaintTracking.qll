private import codeql.dataflow.TaintTracking
private import DataFlowArg
import TaintFlowMake<JSDataFlow, JSTaintFlow>
