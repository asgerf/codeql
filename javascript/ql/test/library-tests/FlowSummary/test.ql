import javascript
import semmle.javascript.dataflow2.DataFlow as DataFlow2
import semmle.javascript.dataflow2.TaintTracking as TaintTracking2
import semmle.javascript.dataflow2.BarrierGuards
import semmle.javascript.dataflow2.FlowSummary
import testUtilities.ConsistencyChecking

DataFlow::CallNode getACall(string name) {
  result.getCalleeName() = name
  or
  result.getCalleeNode().getALocalSource() = DataFlow::globalVarRef(name)
}

module ConfigArg implements DataFlow2::ConfigSig {
  predicate isSource(DataFlow::Node node) { node = getACall("source") }

  predicate isSink(DataFlow::Node node) { node = getACall("sink").getAnArgument() }

  predicate isBarrier(DataFlow::Node node) {
    node.(DataFlow::InvokeNode).getCalleeName().matches("sanitizer_%")
    or
    barrierGuardBlocksNode(node, _)
  }
}

module Configuration = TaintTracking2::Global<ConfigArg>;

class BasicSanitizerGuard extends TaintTracking::SanitizerGuardNode, DataFlow::CallNode {
  BasicSanitizerGuard() { this = getACall("isSafe") }

  override predicate sanitizes(boolean outcome, Expr e) {
    outcome = true and e = this.getArgument(0).asExpr()
  }
}

class ConsistencyConfig extends ConsistencyConfiguration {
  ConsistencyConfig() { this = "ConsistencyConfig" }

  override DataFlow::Node getAnAlert() { Configuration::flow(_, result) }
}

class FlowThrough extends SummarizedCallable {
  FlowThrough() { this = "flowThrough" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowThrough") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0]" and output = "ReturnValue" and preservesValue = true
  }
}

class FlowThroughTaint extends SummarizedCallable {
  FlowThroughTaint() { this = "flowThroughTaint" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowThroughTaint") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0]" and output = "ReturnValue" and preservesValue = false
  }
}

class FlowIntoProp extends SummarizedCallable {
  FlowIntoProp() { this = "flowIntoProp" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowIntoProp") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0]" and output = "ReturnValue.Member[prop]" and preservesValue = true
  }
}

class FlowOutOfProp extends SummarizedCallable {
  FlowOutOfProp() { this = "flowOutOfProp" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowOutOfProp") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0].Member[prop]" and output = "ReturnValue" and preservesValue = true
  }
}

class FlowIntoArrayElement extends SummarizedCallable {
  FlowIntoArrayElement() { this = "flowIntoArrayElement" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowIntoArrayElement") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0]" and output = "ReturnValue.ArrayElement" and preservesValue = true
  }
}

class FlowIntoPromise extends SummarizedCallable {
  FlowIntoPromise() { this = "flowIntoPromise" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowIntoPromise") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0]" and output = "ReturnValue.Awaited" and preservesValue = true
  }
}

class FlowOutOfPromise extends SummarizedCallable {
  FlowOutOfPromise() { this = "flowOutOfPromise" }

  override DataFlow::InvokeNode getACall() { result = getACall("flowOutOfPromise") }

  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
    input = "Argument[0].Awaited" and output = "ReturnValue" and preservesValue = true
  }
}
