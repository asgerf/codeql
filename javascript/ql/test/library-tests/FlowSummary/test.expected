| library-tests/FlowSummary/tst.js:278 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:282 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:283 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:286 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:287 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:290 | expected an alert, but found none | NOT OK | ConsistencyConfig |
| library-tests/FlowSummary/tst.js:291 | expected an alert, but found none | NOT OK | ConsistencyConfig |
